File: getting_started/rest/rest.java
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.core.type.TypeReference;

import java.nio.file.*;
import java.net.URI;
import java.net.http.*;
import java.security.KeyFactory;
import java.security.interfaces.RSAPrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class rest {
    private static final ObjectMapper mapper = new ObjectMapper();
    private static final HttpClient httpClient = HttpClient.newHttpClient();

    public static String createJwt(String serviceAccountPath) throws Exception {
        byte[] jsonData = Files.readAllBytes(Paths.get(serviceAccountPath));
        var keyData = mapper.readTree(jsonData);

        String privateKeyContent = keyData.get("private_key").asText()
                .replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replaceAll("\\s", "");

        byte[] encoded = Base64.getDecoder().decode(privateKeyContent);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(encoded);
        RSAPrivateKey privateKey = (RSAPrivateKey) KeyFactory.getInstance("RSA").generatePrivate(keySpec);

        Algorithm algorithm = Algorithm.RSA256(null, privateKey);
        String email = keyData.get("client_email").asText();
        String aud = email.split("@")[0] + ".roguedb.dev";

        return JWT.create()
                .withKeyId(keyData.get("private_key_id").asText())
                .withIssuer(email)
                .withSubject(email)
                .withAudience(aud)
                .withIssuedAt(Instant.now())
                .withExpiresAt(Instant.now().plusSeconds(3600))
                .sign(algorithm);
    }

    public static List<String> detectFiles(Path dir) throws Exception {
        try (var stream = Files.walk(dir)) {
            return stream
                .filter(p -> p.toString().endsWith(".proto"))
                .map(p -> {
                    try { return Files.readString(p); } 
                    catch (Exception e) { return ""; }
                })
                .collect(Collectors.toList());
        }
    }

    public static void main(String[] args) throws Exception {
    	// See purchase confirmation emails for details and service_account.json.
        String url = "DATABASE_URL";
        String apiKey = "API_KEY";
        String token = createJwt("/path/to/service_account.json");

        ////////////////////////////////////////////////
        //// Insert, Update, and Remove API Example ////
        ////////////////////////////////////////////////

        // Creating an Insert, Update, or Remove API with JSON. 
        ObjectNode requestBody = mapper.createObjectNode();
        requestBody.put("api_key", apiKey);
        var messages = requestBody.putArray("messages");
        var testMsg = messages.addObject();
        
        // @type: After '/', matches proto package and message name
        testMsg.put("@type", "type.googleapis.com/rogue.services.Test");
        // attribute1: Field name in Test
        testMsg.put("attribute1", 10);

        // REST call for Insert API.
        HttpRequest request = HttpRequest.newBuilder()
            .header("Authorization", "Bearer " + token)
            .header("Content-Type", "application/json")
            .uri(URI.create("https://" + url + "/rest/insert"))
            // .uri(URI.create("https://" + url + "/rest/update")) // REST call for Update API.
            // .uri(URI.create("https://" + url + "/rest/remove")) // REST call for Remove API
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            // .PATCH(HttpRequest.BodyPublishers.ofString(requestBody.toString())) // REST call for Update API
            // .DELETE(HttpRequest.BodyPublishers.ofString(requestBody.toString())) // REST call for Remove API
            .build();
        
        // No response given. Errors reported in status code.
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        ///////////////////////////////
        ///// Search API Examples /////
        ///////////////////////////////
        
        // Example of a basic index query. 
        // For Test, attribute1, attribute2, and attribute3 form the index.
        // Search Query: 
        // Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
        // AND
        // Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
        requestBody = mapper.createObjectNode();
        requestBody.put("api_key", apiKey);
        ArrayNode queries = requestBody.putArray("queries");
        ObjectNode query = queries.addObject();
        ObjectNode basic = query.putObject("basic");

        ArrayNode comparisons = basic.putArray("comparisons");
        comparisons.add("GREATER_EQUAL");
        comparisons.add("LESSER_EQUAL");

        ArrayNode operands = basic.putArray("operands");
        ObjectNode operand1 = mapper.createObjectNode();
        
        operand1.put("@type", "type.googleapis.com/rogue.services.Test");
        operand1.put("attribute1", 1);
        operand1.put("attribute2", 1);
        operand1.put("attribute3", true);
        operands.add(operand1);

        ObjectNode operand2 = mapper.createObjectNode();          
        operand2.put("@type", "type.googleapis.com/rogue.services.Test");
        operand2.put("attribute1", 10);
        operand2.put("attribute2", 10);
        operand2.put("attribute3", true);
        operands.add(operand2);

        // All search query types use this URL
        request = HttpRequest.newBuilder()
            .header("Authorization", "Bearer " + token)
            .header("Content-Type", "application/json")
            .uri(URI.create("https://" + url + "/rest/search"))
            .method("GET", HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build();

        var root = mapper.readTree(response.body());
        var results = mapper.convertValue(
            root.get("results"), 
            new TypeReference<Map<String, List<Object>>>() {});;

        // Example of a basic non-indexed query.
        // Search Query: Test.attribute1 < 1 Test.attribute2 != 10
        requestBody.put("api_key", apiKey); // ensure api_key still present
        queries = requestBody.putArray("queries");
        query = queries.addObject();
        basic = query.putObject("basic");

        comparisons = basic.putArray("comparisons");
        comparisons.add("LESSER");
        comparisons.add("NOT_EQUAL");

        ArrayNode fields = basic.putArray("fields");
        fields.add(1);
        fields.add(2);

        operands = basic.putArray("operands");
        operand1 = mapper.createObjectNode();
        operand1.put("@type", "type.googleapis.com/rogue.services.Test");
        operand1.put("attribute1", 1);
        operands.add(operand1);

        operand2 = mapper.createObjectNode();
        operand2.put("@type", "type.googleapis.com/rogue.services.Test");
        operand2.put("attribute2", 10);
        operands.add(operand2);

        /////////////////////////////////////
        ///// Schema Change API Example /////
        /////////////////////////////////////
        
        requestBody = mapper.createObjectNode();
        requestBody.put("api_key", apiKey);
        ArrayNode schemasArray = requestBody.putArray("schemas");
        for(String file : detectFiles(Paths.get("/path/to/protos"))) {
            // Proto file definitions. No modifications required.
            schemasArray.add(Files.readString(Paths.get(file)));
        }

        request = HttpRequest.newBuilder()
            .header("Authorization", "Bearer " + token)
            .header("Content-Type", "application/json")
            .uri(URI.create("https://" + url + "/rest/subscribe"))
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build();

        // Any schemas excluded will have associated data deleted.
        // Schema change failure results in no changes applied.
        response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }
}

File: getting_started/rest/rest.js
// Install deps: npm install axios jsonwebtoken
const fs = require('fs').promises;
const path = require('path');
const axios = require('axios');
const jwt = require('jsonwebtoken');

async function createJwt(serviceAccountPath, expireMinutes = 60) {
    const raw = await fs.readFile(serviceAccountPath, 'utf8');
    const keyData = JSON.parse(raw);

    const now = Math.floor(Date.now() / 1000);
    const payload = {
        iat: now,
        exp: now + expireMinutes * 60,
        iss: keyData.client_email,
        sub: keyData.client_email,
        aud: `${keyData.client_email.split('@')[0]}.roguedb.dev`, };

    // keyid placed in header automatically if provided via options.keyid
    return jwt.sign(payload, keyData.private_key, {
        algorithm: 'RS256',
        keyid: keyData.private_key_id,
        header: { typ: 'JWT' },
        });
}

function detectFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    files.forEach(file => {
        const filePath = path.join(dir, file);
        if (fs.statSync(filePath).isDirectory()) {
            detectFiles(filePath, fileList);
        } else if (path.extname(file) === '.proto') {
            fileList.push(filePath);
        }
    });
    return fileList;
}

async function main() {
    // See purchase confirmation emails for details and service_account.json.
    const url = 'DATABASE_URL';
    const apiKey = 'API_KEY';
    const jwtPath = '/path/to/service_account.json';

    const token = await createJwt(jwtPath);
    const headers = {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
    };

    ////////////////////////////////////////////////////////
    ///////  Insert, Update, and Remove API Example  ///////
    ////////////////////////////////////////////////////////
  
    const requestBody = {
        api_key: apiKey,
        messages: [
            {
            // @type: After '/', matches proto package and message name
            '@type': 'type.googleapis.com/rogue.services.Test',
            // attribute1: Field name in Test
            attribute1: 10,
            },],};
    const response = await axios.post(`https://${url}/rest/insert`, requestBody, { headers }); // Insert API
    // const response = await axios.patch(`https://${url}/rest/update`, requestBody, { headers }); // Update API
    // const response = await axios.delete(`https://${url}/rest/remove`, { headers, data: requestBody }); // Remove API

    //////////////////////////////////////
    ////////  Search API Example  ////////
    //////////////////////////////////////

    // Example of a basic index query. 
    // For Test, attribute1, attribute2, and attribute3 form the index.
    // Search Query: 
    // Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
    // AND
    // Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
    const searchIndex = {
        api_key: apiKey,
        queries: [
        {
            basic: {
                comparisons: ['GREATER_EQUAL', 'LESSER_EQUAL'],
                operands: [
                {
                    '@type': 'type.googleapis.com/rogue.services.Test',
                    attribute1: 1,
                    attribute2: 1,
                    attribute3: true,
                },
                {
                    '@type': 'type.googleapis.com/rogue.services.Test',
                    attribute1: 10,
                    attribute2: 10,
                    attribute3: true,
                },],},},],};

    // All search query types use this URL
    const searchResp = await axios.post(`https://${url}/rest/search`, searchIndex, { headers });

    // Queries are zero-indexed. 
    // Results are mapped to results field.
    // All messages are stored in the messages field.
    const firstResults = searchResp.data?.results?.[0] ?? [];
    for (const r of firstResults) {
        continue;
    }

    // Example of a basic non-indexed query.
    // Search Query: Test.attribute1 < 1 and Test.attribute2 != 10
    const searchNonIndexed = {
        api_key: apiKey,
        queries: [
            {
                basic: {
                    comparisons: ['LESSER', 'NOT_EQUAL'],
                    fields: [1, 2], // Corresponds to attribute1 and attribute2 field ids in Test
                    operands: [
                        { '@type': 'type.googleapis.com/rogue.services.Test', attribute1: 1 },
                        { '@type': 'type.googleapis.com/rogue.services.Test', attribute2: 10 },
                    ],},},],};

    ///////////////////////////////////
    //// Schema Change API Example ////
    ///////////////////////////////////

    const protoFiles = await detectFiles(protoDirs);
    const schemas = [];
    detectFiles('/path/to/protos').forEach(file => {
        // All proto files should be sent in a list of
        // their contents. No modifications required.
        schemas.push(fs.readFileSync(file, 'utf8'));
    });
  
    // Any schemas excluded will have associated data deleted.
    // Schema change failure results in no changes applied.
    const subResp = await axios.post(
      `https://${url}/rest/subscribe`,
      { api_key: apiKey, schemas },
      { headers }
    );
}

main()